src/Processors/QueryPlan/ReadFromMergeTree.cpp

```diff
@@ -3300,7 +3300,7 @@ void ReadFromMergeTree::replaceColumnsForTextSearch(const IndexReadColumns & add
 
 ConditionSelectivityEstimatorPtr ReadFromMergeTree::getConditionSelectivityEstimator() const
 {
-    return data.getConditionSelectivityEstimatorByPredicate(storage_snapshot, nullptr, getContext());
```

Comment:
Maybe it makes sense to cache the estimator in ReadFromMergeTree, because now we may build it twice: in prewhere optimization and in joins optimization.

Reply:
Agreed
maybe also cache it in `MergeTreeData`